starknet/src/utils/types.cairo

```diff
@@ -6,14 +6,88 @@ use hash::LegacyHash;
 use option::OptionTrait;
 use clone::Clone;
 use starknet::{
-    ContractAddress, Store, contract_address_const, StorageBaseAddress, SyscallResult,
+    EthAddress, ContractAddress, Store, contract_address_const, StorageBaseAddress, SyscallResult,
     storage_write_syscall, storage_read_syscall, storage_address_from_base_and_offset,
     storage_base_address_from_felt252, contract_address::Felt252TryIntoContractAddress,
     syscalls::deploy_syscall, class_hash::Felt252TryIntoClassHash
 };
 use sx::utils::math::pow;
 use array::SpanTrait;
 
+#[derive(Copy, Drop, Serde, LegacyHash, PartialEq, starknet::Store)]
+enum UserAddress {
+    // Starknet address type
+    StarknetAddress: ContractAddress,
+    // Ethereum address type
+    EthereumAddress: EthAddress,
+    // Custom address type to provide compatibility with any address that can be represented as a u256.
+    CustomAddress: u256
+}
+
+impl LegacyHashUserAddress of LegacyHash<UserAddress> {
+    fn hash(state: felt252, value: UserAddress) -> felt252 {
+        match value {
+            UserAddress::StarknetAddress(address) => LegacyHash::<felt252>::hash(
+                state, address.into()
+            ),
+            UserAddress::EthereumAddress(address) => LegacyHash::<felt252>::hash(
+                state, address.into()
+            ),
+            UserAddress::CustomAddress(address) => LegacyHash::<u256>::hash(state, address),
+        }
+    }
+}
+
+trait UserAddressTrait {
+    fn to_starknet_address(self: UserAddress) -> ContractAddress;
+    fn to_ethereum_address(self: UserAddress) -> EthAddress;
+    fn to_custom_address(self: UserAddress) -> u256;
+}
+
+impl UserAddressImpl of UserAddressTrait {
+    fn to_starknet_address(self: UserAddress) -> ContractAddress {
+        match self {
+            UserAddress::StarknetAddress(address) => address,
```

Comment:
in Rust you can do :
```
match self {
  UserAddress::StarknetAddress(address) => address,
  _ => panic_with_felt252('Incorrect address type'),
}
```
does it work in cairo?

Reply:
i dont think so yet. https://book.cairo-lang.org/ch05-02-the-match-control-flow-construct.html#match-0-and-the-_-placeholder

Reply:
`Unsupported match. Currently, matches require one arm per variant, in the order of variant definition.`

Reply:
sucks, weirt that the cairo book doesn't mention that. Will probs come in the future